x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/AuthorizationEngine.java

```diff
@@ -667,6 +697,103 @@ public IndicesAccessControl getIndicesAccessControl() {
+            "authorization",
+            true,
+            (a) -> new ParentActionAuthorization((String) a[0])
+        );
+        static {
+            PARSER.declareString(constructorArg(), ACTION);
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            out.writeString(action);
+        }
+
+        @Override
+        public XContentBuilder toXContent(XContentBuilder builder, ToXContent.Params params) throws IOException {
+            builder.field(ACTION.getPreferredName(), action);
+            return builder;
+        }
+
+        /**
+         * Reads an {@link ParentActionAuthorization} from a {@link StreamInput}
+         *
+         * @param in the {@link StreamInput} to read from
+         * @return {@link ParentActionAuthorization}
+         * @throws IOException if I/O operation fails
+         */
+        public static ParentActionAuthorization readFrom(StreamInput in) throws IOException {
+            String action = in.readString();
+            return new ParentActionAuthorization(action);
+        }
+
+        /**
+         * Reads {@link ParentActionAuthorization} from {@link XContentParser}
+         *
+         * @param parser {@link XContentParser}
+         * @return {@link ParentActionAuthorization}
+         * @throws IOException if I/O operation fails
+         */
+        public static ParentActionAuthorization fromXContent(final XContentParser parser) throws IOException {
+            return PARSER.apply(parser, null);
+        }
+
+        /**
+         * Read and deserialize parent authorization from thread context.
+         *
+         * @param context the thread context to read from
+         * @return {@link ParentActionAuthorization} or null
+         * @throws IOException if reading fails due to I/O exception
+         */
+        @Nullable
+        public static ParentActionAuthorization readFromThreadContext(ThreadContext context) throws IOException {
+            final String header = context.getHeader(THREAD_CONTEXT_KEY);
+            if (header == null) {
+                return null;
+            }
+
+            byte[] bytes = Base64.getDecoder().decode(header);
+            StreamInput input = StreamInput.wrap(bytes);
+            return readFrom(input);
+        }
+
+        /**
+         * Writes the authorization to the context. There must not be an existing authorization in the context and if there is an
+         * {@link IllegalStateException} will be thrown.
+         */
+        public void writeToThreadContext(ThreadContext context) throws IOException {
+            ensureContextDoesNotContainAuthorization(context);
+            String header = this.encode();
+            assert header != null : "parent authorization object encoded to null";
+            context.putHeader(THREAD_CONTEXT_KEY, header);
+        }
+
+        private void ensureContextDoesNotContainAuthorization(ThreadContext context) {
```

Comment:
nit: it looks like the `ThreadContext#putHeader` already has checks to not overwrite headers, so this check looks redundant.

Reply:
Good point. I've removed the check.